migtools/oadp-non-admin
# Reconcile loop rework

NonAdminBackup reconcile loop major rework
    
Rework of the reconcile loop to include batch reconcile.
    
Move assignment of the Log or Context outside of the reconcile loop.

# How to test

Note it's best to test this PR together with https://github.com/migtools/oadp-non-admin/pull/56 which includes namespace guard (case d) in this instructions.

1. Setup non-admin user as per instructions: https://github.com/migtools/oadp-non-admin/blob/master/docs/non_admin_user.md
2. Create multiple NonAdminBackup objects in the namespace for which the non-admin user have namespace admin rights. After each creation see if at the end of reconcile the NonAdminBackup is correctly updated with its status and if in some cases VeleroBackup object was created in the openshift-adp namespace (as an admin).

All of the below are in the non-admin namespace `nacproject`:

a) Case where there is no backupSpec:

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: example-nospec
  namespace: nacproject
spec: {}
```

Result should be NAB in backingOff phase with `BackupSpec is not defined` condition message:

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: example-nospec
  namespace: nacproject
spec: {}
status:
  conditions:
    - lastTransitionTime: '2024-05-02T08:24:50Z'
      message: BackupSpec is not defined
      reason: InvalidBackupSpec
      status: 'False'
      type: Accepted
  phase: BackingOff

```


b) Case where there is backupSpec, but empty, should create Velero Backup with nacproject namespace in the includedNamespaces (below is not full spec, removed `managedFields` from it):

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: empty-backupspec
  namespace: nacproject
spec: 
  backupSpec: {}
```

Result should be created Velero Backup with `includedNamespaces` same as the origin NAB namespace and updated NonAdminBackup (below is not full spec, removed `managedFields` from it):

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  creationTimestamp: '2024-05-02T08:27:46Z'
  name: empty-backupspec
  namespace: nacproject
  resourceVersion: '71261871'
  uid: 624c270e-2711-490d-923e-aaebfcc2a420
spec:
  backupSpec:
    volumeSnapshotLocations:
      - velero-sample-1
    defaultVolumesToFsBackup: false
    csiSnapshotTimeout: 10m0s
    ttl: 720h0m0s
    itemOperationTimeout: 4h0m0s
    metadata: {}
    storageLocation: velero-sample-1
    hooks: {}
    includedNamespaces:
      - nacproject
    snapshotMoveData: false
status:
  conditions:
    - lastTransitionTime: '2024-05-02T08:27:47Z'
      message: backup accepted
      reason: BackupAccepted
      status: 'True'
      type: Accepted
    - lastTransitionTime: '2024-05-02T08:27:57Z'
      message: Created Velero Backup object
      reason: BackupScheduled
      status: 'True'
      type: Queued
  phase: Created
  veleroBackupName: nab-nacproject-d0b80d478b6a26
  veleroBackupNamespace: openshift-adp
  veleroBackupStatus:
    expiration: '2024-06-01T08:27:57Z'
    failureReason: >-
      unable to get credentials: unable to get key for secret: Secret
      "cloud-credentials" not found
    formatVersion: 1.1.0
    phase: Failed
    startTimestamp: '2024-05-02T08:27:57Z'
    version: 1
```

c) Perform step a) and then update/patch (a) to include empty backupSpec. The reconcile should be same as in b), so the Velero Backup should be created.

d) Create NonAdminBackup with `includedNamespaces` that does not match the namespace from the NonAdminBackup.

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  name: example-wrongnamespace
  namespace: nacproject
spec:
   backupSpec:
     includedNamespaces:
       - othernamespace
```

Result should be NAB in backingOff phase with `spec.backupSpec.IncludedNamespaces can not contain namespaces other then: nacproject` condition message:

```
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  creationTimestamp: '2024-05-02T08:33:35Z'
  generation: 1
  name: example-wrongnamespace
  namespace: nacproject
  resourceVersion: '71264312'
  uid: 2716a72d-3e44-466d-8126-5bcdeed9413b
spec:
  backupSpec:
    includedNamespaces:
      - othernamespace
status:
  conditions:
    - lastTransitionTime: '2024-05-02T08:33:45Z'
      message: >-
        spec.backupSpec.IncludedNamespaces can not contain namespaces other
        then: nacproject
      reason: InvalidBackupSpec
      status: 'False'
      type: Accepted
  phase: BackingOff
```
